url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/engineio/client/Socket.java

```diff
@@ -0,0 +1,779 @@
+                } catch (JSONException e) {
+                    this.emit(EVENT_ERROR, new EngineIOException(e));
+                }
+            } else if (Packet.PONG.equals(packet.type)) {
+                this.setPing();
+            } else if (Packet.ERROR.equals(packet.type)) {
+                EngineIOException err = new EngineIOException("server error");
+                err.code = packet.data;
+                this.emit(EVENT_ERROR, err);
+            } else if (Packet.MESSAGE.equals(packet.type)) {
+                this.emit(EVENT_DATA, packet.data);
+                this.emit(EVENT_MESSAGE, packet.data);
+            }
+        } else {
+            logger.fine(String.format("packet received with socket readyState '%s'", this.readyState));
+        }
+    }
+
+    private void onHandshake(HandshakeData data) {
+        this.emit(EVENT_HANDSHAKE, data);
+        this.id = data.sid;
+        this.transport.query.put("sid", data.sid);
+        this.upgrades = this.filterUpgrades(Arrays.asList(data.upgrades));
+        this.pingInterval = data.pingInterval;
+        this.pingTimeout = data.pingTimeout;
+        this.onOpen();
+        // In case open handler closes socket
+        if (ReadyState.CLOSED == this.readyState) return;
+        this.setPing();
+
+        this.off(EVENT_HEARTBEAT, this.onHeartbeatAsListener);
+        this.on(EVENT_HEARTBEAT, this.onHeartbeatAsListener);
+    }
+
+    private final Listener onHeartbeatAsListener = new Listener() {
+        @Override
+        public void call(Object... args) {
+            Socket.this.onHeartbeat(args.length > 0 ? (Long)args[0]: 0);
+        }
+    };
+
+    private void onHeartbeat(long timeout) {
+        if (this.pingTimeoutTimer != null) {
+            pingTimeoutTimer.cancel(true);
+        }
+
+        if (timeout <= 0) {
+            timeout = this.pingInterval + this.pingTimeout;
+        }
+
+        final Socket self = this;
+        this.pingTimeoutTimer = this.heartbeatScheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (self.readyState == ReadyState.CLOSED) return;
+                        self.onClose("ping timeout");
+                    }
+                });
+            }
+        }, timeout, TimeUnit.MILLISECONDS);
+    }
+
+    private void setPing() {
+        if (this.pingIntervalTimer != null) {
+            pingIntervalTimer.cancel(true);
+        }
+
+        final Socket self = this;
+        this.pingIntervalTimer = this.heartbeatScheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        logger.fine(String.format("writing ping packet - expecting pong within %sms", self.pingTimeout));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format long: self.pingTimeout. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 531. Consider formatting this data with &#39;%d&#39; instead.